pkg/controllers/mcs/service_export_controller.go

```diff
@@ -0,0 +1,493 @@
+			AddFunc: func(obj interface{}, cluster string) {
+				eps := obj.(*discoveryv1b1.EndpointSlice)
+				c.worker.Enqueue(reconcileKey{
+					cluster:   cluster,
+					gvk:       common.EndpointSliceGVK,
+					namespace: eps.Namespace,
+					name:      eps.Name,
+				})
+			},
+			UpdateFunc: func(oldObj, newObj interface{}, cluster string) {
+				eps := newObj.(*discoveryv1b1.EndpointSlice)
+				c.worker.Enqueue(reconcileKey{
+					cluster:   cluster,
+					gvk:       common.EndpointSliceGVK,
+					namespace: eps.Namespace,
+					name:      eps.Name,
+				})
+			},
+			DeleteFunc: func(obj interface{}, cluster string) {
+				if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+					obj = tombstone.Obj
+					if obj == nil {
+						return
+					}
+				}
+				eps := obj.(*discoveryv1b1.EndpointSlice)
+				c.worker.Enqueue(reconcileKey{
+					cluster:   cluster,
+					gvk:       common.EndpointSliceGVK,
+					namespace: eps.Namespace,
+					name:      eps.Name,
+				})
+			},
+		},
+	})
+
+	return c, nil
+}
+
+func (c *ServiceExportController) Run(ctx context.Context) {
+	ctx, logger := logging.InjectLogger(ctx, c.logger)
+
+	logger.Info("Starting controller")
+	defer logger.Info("Stopping controller")
+
+	go c.endpointSliceInformer.Informer().Run(ctx.Done())
+
+	if !cache.WaitForNamedCacheSync(ServiceExportControllerName, ctx.Done(), c.HasSynced) {
+		logger.Error(nil, "Timed out waiting for caches to sync")
+		return
+	}
+	logger.Info("Caches are synced")
+	c.worker.Run(ctx)
+	<-ctx.Done()
+}
+
+func (c *ServiceExportController) reconcile(ctx context.Context, reconcileKey reconcileKey) (status worker.Result) {
+	key := reconcileKey.String()
+	ctx, keyedLogger := logging.InjectLoggerValues(ctx, "control-loop", "reconcile", "object", key)
+
+	switch reconcileKey.gvk {
+	case common.ServiceExportGVK:
+		if err := c.reconcileServiceExport(ctx, reconcileKey); err != nil {
+			keyedLogger.Error(err, fmt.Sprintf("failed to reconcile serviceExport(%s/%s)", reconcileKey.namespace, reconcileKey.name))
+			return worker.StatusError
+		}
+	case common.EndpointSliceGVK:
+		if err := c.reconcileEndpointSlice(ctx, reconcileKey); err != nil {
+			keyedLogger.Error(err, fmt.Sprintf("failed to reconcile endpointSlice(%s/%s)", reconcileKey.namespace, reconcileKey.name))
+			return worker.StatusError
+		}
+	}
+
+	return worker.StatusAllOK
+}
+
+func (c *ServiceExportController) reconcileServiceExport(ctx context.Context, key reconcileKey) error {
+	logger := klog.FromContext(ctx)
+
+	needReconcile, err := c.filterServiceExport(ctx, key)
+	if err != nil {
+		return fmt.Errorf("failed to check if serviceExport needs to be filtered: %w", err)
+	}
+	if !needReconcile {
+		return nil
+	}
+
+	seLister, synced, exists := c.fedInformerManager.GetServiceExportLister(key.cluster)
+	if !exists {
+		return fmt.Errorf("informer of serviceExport not exists for cluster %s", key.cluster)
+	}
+	if !cache.WaitForCacheSync(ctx.Done(), synced) {
```

Comment:
Will this ever timeout? More importantly, this goroutine will be occupied until this function returns. Potentially all available `ReconcileWorker` goroutines might be stuck here and new events will not be processed.

Reply:
Got it. I had not discovered this hidden problem before.